sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -2652,5 +2631,169 @@ def my_pipeline(x: Input[Artifact] = Artifact(
+        ):
+
+            @dsl.pipeline
+            def math_pipeline() -> int:
+                with dsl.ParallelFor([1, 2, 3]) as f:
+                    t = double(num=f)
+
+                return add(nums=t.output).output
+
+    def test_producer_condition_legal1(self):
+        from kfp import dsl
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        @dsl.pipeline
+        def math_pipeline(text: str) -> int:
+            with dsl.Condition(text == 'text'):
+                with dsl.ParallelFor([1, 2, 3]) as f:
+                    t = double(num=f)
+
+                return add(nums=dsl.Collected(t.output)).output
+
+    def test_producer_condition_legal2(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        @dsl.pipeline
+        def my_pipeline(text: str = ''):
+            with dsl.ParallelFor([1, 2, 3]) as f:
+                with dsl.Condition(text == 'text'):
+                    t = double(num=f)
+
+            with dsl.Condition(text == 'text'):
+                x = add(nums=dsl.Collected(t.output))
+
+    def test_producer_condition_illegal1(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        with self.assertRaisesRegex(
+                compiler_utils.InvalidTopologyException,
+                r'Tasks cannot depend on an upstream task inside Condition that is not a common ancestor of both tasks\. Task add depends on upstream task double\.'
+        ):
+
+            @dsl.pipeline
+            def my_pipeline(text: str = ''):
+                with dsl.Condition(text == 'text'):
+                    with dsl.ParallelFor([1, 2, 3]) as f:
+                        t = double(num=f)
+
+                with dsl.Condition(text == 'text'):
+                    x = add(nums=dsl.Collected(t.output))
+
+    def test_producer_condition_illegal2(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        with self.assertRaisesRegex(
+                compiler_utils.InvalidTopologyException,
+                r'Tasks cannot depend on an upstream task inside Condition that is not a common ancestor of both tasks\. Task add depends on upstream task double\.'
+        ):
+
+            @dsl.pipeline
+            def my_pipeline(text: str = ''):
+                with dsl.Condition(text == 'a'):
+                    with dsl.Condition(text == 'b'):
```

Comment:
The second condition would never hit. Again, we should be realistic on the sample code, even if it's the invalid case--it should have one and only one "error", but not multiple mistakes which could distract the focus.  
Secondly, given the indentation level of `add`, the first condition actually has nothing to do with the error, yet it could be a red herring.